docs/notebooks/How_JAX_primitives_work.ipynb

```diff
@@ -0,0 +1,1638 @@
+        "def sq_add_numpy(a, b):\n",
+        "    return ma_numpy(a, a, b)\n",
+        "\n",
+        "print(\"\\nNormal evaluation:\")  \n",
+        "print(\"sq_add_numpy = \", sq_add_numpy(2., 10.))\n",
+        "print(\"\\nGradient evaluation:\")\n",
+        "print(\"grad(sq_add_numpy) = \", api.grad(sq_add_numpy)(2.0, 10.))"
+      ],
+      "execution_count": 0,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "text": [
+            "\n",
+            "Normal evaluation:\n",
+            "call sq_add_numpy(2.0, 10.0)\n",
+            "  call ma_numpy(2.0, 2.0, 10.0)\n",
+            "  |<- ma_numpy = 14.0\n",
+            "|<- sq_add_numpy = 14.0\n",
+            "sq_add_numpy =  14.0\n",
+            "\n",
+            "Gradient evaluation:\n",
+            "call sq_add_numpy(Traced<ConcreteArray(2.0)>, 10.0)\n",
+            "  call ma_numpy(Traced<ConcreteArray(2.0)>, Traced<ConcreteArray(2.0)>, 10.0)\n",
+            "  |<- ma_numpy = Traced<ConcreteArray(14.0)>\n",
+            "|<- sq_add_numpy = Traced<ConcreteArray(14.0)>\n",
+            "grad(sq_add_numpy) =  4.0\n"
+          ],
+          "name": "stdout"
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "Sg-D8EdeFn4a",
+        "colab_type": "text"
+      },
+      "source": [
+        "Notice that in the process of computing `grad`, JAX invokes `sq_add_numpy` and\n",
+        "`ma_numpy` with special arguments `ConcreteArray(...)` (described further \n",
+        "below in this colab). \n",
+        "It is important to remember that a JAX-traceable function must be able to \n",
+        "operate not only on concrete arguments but also on special abstract arguments\n",
+        "that JAX may use to abstract the function execution.\n",
+        "\n",
+        "The JAX traceability property is satisfied as long as the function is written \n",
+        "in terms of JAX primitives. "
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "WxrQO7-XGLcg",
+        "colab_type": "text"
+      },
+      "source": [
+        "# Defining new JAX primitives\n",
+        "\n",
+        "The right way to add support for multiple-add is in terms of existing\n",
+        "JAX primitives, as shown above. However, in order to demonstrate how JAX\n",
+        "primitives work let us pretend that we want to add a new primitive to \n",
+        "JAX for the multiply-add functionality."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "cPqAH1XOGTN4",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "source": [
+        "from jax import core\n",
+        "ma_p = core.Primitive(\"ma\")  # Create the primitive\n",
+        "\n",
+        "@trace(\"ma_prim\")\n",
+        "def ma_prim(x, y, z):\n",
+        "  \"\"\"The JAX-traceable way to use the JAX primitive.\"\"\"\n",
+        "  return ma_p.bind(x, y, z)\n",
```

Comment:
This might be a good place to mention that possibly traced variables must always be passed by position directly into `.bind()` vs. non-traced variables that can be passed as keyword arguments. This means that if you want to accept structured arguments, you'll need to flatten them (with `jax.tree_util`) before calling the primitive itself.

Reply:
Added a clarification.